datasimulator/graph.py

```diff
@@ -0,0 +1,265 @@
+                            if not isinstance(sub_links, list):
+                                sub_links = [sub_links]
+
+                            # just pick one of sub-group links
+                            for sub_link in sub_links:
+                                if "target_type" in sub_link:
+                                    self._add_required_link_to_node(
+                                        node,
+                                        sub_link["target_type"],
+                                        sub_link.get("name"),
+                                        sub_link.get("multiplicity"),
+                                    )
+                                    break
+
+            except TypeError as e:
+                raise DictionaryError(
+                    "Node {} have non-list links. Detail {}".format(
+                        node.name, e.message
+                    )
+                )
+
+    def generate_submission_order_path_to_node(self, node, submission_order):
+        """
+        Generate submission order so that the current node can be submitted
+
+        Args:
+            node(Node): current node object
+            submission_order(list): submission order list
+
+        Outputs:
+            None
+
+        Side effects:
+            submission_order is updated interatively
+
+        """
+        if node in submission_order:
+            return
+        for linked_node_dict in node.required_links:
+            if linked_node_dict["node"] not in submission_order:
+                self.generate_submission_order_path_to_node(
+                    linked_node_dict["node"], submission_order
+                )
+        submission_order.append(node)
+
+    def generate_submission_order(self, submission_order):
+        """
+        Generate submission order for the graph
+
+        Args:
+            submission_order(list): submission order list
+
+        Outputs:
+            None
+
+        Side effects:
+            submission_order is updated interatively
+        """
+        for node in self.nodes:
+            if node not in submission_order:
+                path_order = []
+                self.generate_submission_order_path_to_node(node, path_order)
+                for item in path_order:
+                    if item not in submission_order:
+                        submission_order.append(item)
+
+    def simulate_graph_data(
+        self, path, n_samples=1, random=True, required_only=True, skip=True
+    ):
+        """
+        Simulate data for the whole graph.
+
+        Args:
+            random(bool): whether randomly link to parent nodes
+            required_only(bool): only simulate required properties
+            skip(bool): skip raising an exception to terminate
+
+        Outputs:
+            None
+        """
+        submission_order = []
+        self.generate_submission_order(submission_order)
```

Comment:
prefer non-side-effectful:
```suggestion
        submission_order = self.generate_submission_order()
```
or just inline that function right here (probably best), since it's only used once here